src/analyzer/analyzerbeats.cpp

```diff
@@ -199,12 +201,34 @@ bool AnalyzerBeats::processSamples(const CSAMPLE* pIn, SINT count) {
         return false;
     }
 
-    m_currentFrame += count / mixxx::kAnalysisChannels;
+    SINT numFrames = count / m_channelCount;
+    const CSAMPLE* pBeatInput = pIn;
+    CSAMPLE* pDrumChannel = nullptr;
+
+    if (m_channelCount > mixxx::kAnalysisChannels) {
+        // If we have multi channel file (a stem file), we extract the first
+        // stereo channel, as it is the drum stem by convention
```

Comment:
I am hesitating to us this assumption unconditionally without a fall back. 
How about check for `StemInfo::getLabel() == "Drums"` and use otherwise the Stereo track?

Reply:
AFAIK this assumption is true for all NI STEMs files, and will be most likely true for DEMUCs based real time stem separation as in stemgen. But for other stem and multichannel formats it might be different.
The label is just a human readable information here, wich is not suitable for compare here, because any string is allowed here.

Reply:
Yeah we cannot compare the strings as it may have an arbitrary value. Reading the spec, the first stem must always be the rhythmic stem, but I guess I could add a setting in the preferences to allow using all stem instead, if anyone was using some non compliant stem file, wdyt?

Reply:
If we are convinced that this assumption is always true. The code is just fine. 
A preference option will not really help if there is a single track breaking this assumption.

Maybe a Todi summarizing our discussion is just fine.

Reply:
I did a  quick search at GitHub and found that almost all implementations call the first stem "Drums" following the NI speck closely. 

But I have also other implementations 

"Bass":
https://github.com/stemrollerapp/stemroller/blob/9a397d21e66d28b3adfc5f13b31598ac3a7ff171/main-src/processQueue.cjs#L265

"Melodia":
https://github.com/underson14/criador-de-stems/blob/master/StemMetadata.json

"Vocals"
https://github.com/jim79/stemgen/blob/dfd451910caf87084cf907351202b754d7c59ff0/4stems-22kHz.json#L6

Not sure if that all affects the NI Stem format. But in these cases it is better to use the full down-mix. 

Since the full down Mixxx already produces reasonable results. We should use it to a to avoid blaming our beat or key detector, just because a user has creative ideas how to sort the stem files.  

What do you think? 


By the way, I have found this testing dataset in NI stem format:
https://sigsep.github.io/datasets/musdb.html

Reply:
This is a label, a text which is only there to be printed in the UI. It is not intended to be machine readable! Each STEM file uses different texts here - and this is intended by the specification.
The machine readable information is the number of the stem channel. According to the specification, the 1st must be the stem with the rythm, and the 4th the vocal stem (if the track has vocals otherwise it should be used for a melodic stem).

We shouldn't try to make assumptions how somebody labels his stem channels. Instead we should strictly follow the specification. And this means we pass the stem labels to the UI only and use the stem channel number for the analyzer.

Reply:
> This is a label, a text which is only there to be printed in the UI.

I can unfortunately not confirm this from the standard. _whenever possible_ a label from the given list shall be used. Since this is not a translated, it does not well suite for the GUI. I think we should make them translatable, this way, additional user Stem labels do not look misplaced. 

And than we have the Arrangement paragraph which is also a weak one compared to the bold must and must not in other parts. _file layout should be consistent_

I assume we will face these tracks:
1. Correct Layout and correct labels -> Use "Drums" 
2. Incorrect Layout and Labels -> Don't use fist for Drums
3. Correct Layout and custom labels -> Don't use fist for Drums
4. Wrong Layout and correct labels -> Use "Drums"

Is a pity withb 3., but we cannot distinguish it from 2. So not using Drums is a save route. 

We can clearly blame the track 4. As broken. User has to fix it and not lable a track "Drums" that don't contain them.


> We shouldn't try to make assumptions how somebody labels his stem channels. 

I agree. But in the same way we should not make assumptions how the track is arranged. Only if the names are correct and we find "Drums" we should assume that it is actually Drums. 

Please also keep in mind that Mixxx is able to analyze tracks with s full stem mix. It has hard times to analyze a bpm of a track where kick drums have been removed. So we must not apply the trick for a better detection when when are not certain that it will help.

Reply:
> we should not make assumptions how the track is arranged

We don't - the spec does.

 > Only if the names are correct and we find "Drums" we should assume that it is actually Drums

Here is the [list of recommended label](https://github.com/acolombier/stemgen?tab=readme-ov-file#note-on-stem-customisation) as per the NI spec - It's not just "drums" that can be rhythmic stem. Also, what about internationalisation? Would "Tambours" or "Schlagzeug" valid as well? How would we handle a Stem metadata with German labels but is loaded on a French interface?  Where would we draw the line?

It really sounds like we are going toward a dead end here. As I said above, I'd be happy to put a user setting to allow ignoring the stem analyse optimisation, so if we ever have a user not respecting the spec, they can still analyse the audio like it would for a traditional stereo for any stem files they load.

Reply:
> I'd be happy to put a user setting to allow ignoring the stem analyse optimisation

I fully support this, using labels for anything ese the GUI output will fail!
And at the same time the DJ media praise the competition for the huge step forward due to AI/STEM base beatgridding. This feature could be the foundation for this.

Reply:
Let's not block this PR over this discussion. 
A logical step for me would be to merge this without the optimization and use always the full down mix. The sups for the optimization can remain. What do you think?

Reply:
> Would "Tambours" or "Schlagzeug" valid as well?

I would argue that translated texts doe not follow the recommendation from the speck. English language is not explicit mentioned, but when you use on one hand one of the preferred terms and than add a German term for a term that is missing in the list it will look misplaced. Using "Tambours" or "Schlagzeug" obviously not allowed because "Drums" the equivalent on the list of recommended terms.  

IMHO the remaining question is how to distinguish case 2. from 3. (from https://github.com/mixxxdj/mixxx/pull/13106#discussion_r1665645429) 
Do we have a real live example for 3.? I have not yet found some. 

>  using labels for anything else the GUI output will fail!

@JoergAtGithub I agree to this statement stand alone. But we have also the evidence that relying on the arrangement also fails which is IMHO worse. 

The degree of failure is different: 
* When the arrangement is wrong with optimization applied the whole analysis fails. This happens for all 2. tracks. 
* When we do not apply analysis because of custom labels the analyse still works, but might be by a chance less accurate. This happens for 3. tracks. 

My conclusion is that we should only apply optimization if we have a sanity check in place to sort out case 2. 

As a workaround we may inform users whether optimizations has been auto applied and allow to force a method using the track context menu. As a foundation we need to store the method in the analysis metadata. 

What do you think?

Reply:
We can not check if the arrangement is correct, but we could say, if the first channel is labled different than expected, that we fall back to mixxed stems.
We currently have already different analyzer options exposed to the user for fixed and variable beatgrid.
Regarding the case that there is a section, where the drum is not playing, I wonder if should only use this stem, or if we maybe should mix the stems with different gains. We could amplify the 1st stem with gain 2 and the 4th with gain 0.5 for example.

Reply:
> the first channel is labled different than expected, that we fall back to mixxed stems.

Do you mean if the label doesn't contain [a recommended label](https://github.com/acolombier/stemgen?tab=readme-ov-file#note-on-stem-customisation)? I guess it could be a valid trade-off, although I would reiterate on having that configurable with a user setting (something in the Analyser like `Optimise analyse for stem track` and 3 options like `Enforced` (always on, like currently), `Automatic` (if the stem is labelled with recommended one - not just `Drums` is other label can be used to indicate rhythmic channel) and `Disabled` (only used mixed stems)

> I wonder if should only use this stem, or if we maybe should mix the stems with different gains

Indeed, there could be some macros optimisation to do here, but this will require further benchmarking to find the ideal parameters. I think the current state of this PR doesn't prevent adding this in a future iteration, and I would rather do it like this as I cannot guarantee I will have the time to do that benchmarking and correct tweaking.

Reply:
> Do you mean if the label doesn't contain [a recommended label](https://github.com/acolombier/stemgen?tab=readme-ov-file#note-on-stem-customisation)? 

Yes. 


>  that configurable with a user setting (something in the Analyser like Optimise analyse for stem track and 3 options like Enforced (always on, like currently), Automatic (if the stem is labelled with recommended one - not just Drums is other label can be used to indicate rhythmic channel) and Disabled (only used mixed stems)

That works for me. Do you agree to have "Disabled" in this PR and do the configuration in a follow up? 

The proposed options are IMHO however to difficult to use. How about using only Automatic but keep an extensible list of labels that will enable the optimization? 

In addition we should always offer a reanalyze context menu, because there are likely some tracks that follow the stem specs but have worse results with optimization enabled.

Reply:
> Do you agree to have "Disabled" in this PR and do the configuration in a follow up?

I have added a basic configuration, which will be `Disabled` by default, and support `Enforced`

> The proposed options are IMHO however to difficult to use. How about using only Automatic but keep an extensible list of labels that will enable the optimization?

That sounds like a potential plan for the future. I have left `Automatic` unimplemented for now so we can revisit and see if there is a value in implementing such a feature 

> In addition we should always offer a reanalyze context menu, because there are likely some tracks that follow the stem specs but have worse results with optimization enabled.

This can be implemented in the future, if we really see a need for it. Currently, one can always:
- Change the analyser strategy in setting
- Re analyse

Agreed this is 6 clicks vs 2 clicks, but I'm not sure the ROI is worth focusing on.

Reply:
Captured as part of #13466